libs/base/System/Concurrency.idr

```diff
@@ -208,6 +363,68 @@ export
 channelGet : HasIO io => (chan : Channel a) -> io a
 channelGet chan = primIO (prim__channelGet chan)
 
+||| Non-blocking version of channelGet.
+|||
+||| @ chan the channel to receive on
+partial
+export
+channelGetNonBlocking : HasIO io => Scheme a => (chan : Channel a) -> io (Maybe a)
+channelGetNonBlocking chan =
+  pure $ (fromScheme . decodeObj) !(primIO (prim__channelGetNonBlocking chan))
+  where
+    decodeObj : ChannelObj -> ChannelSchemeObj
+    decodeObj obj =
+      if prim_isInteger obj == 1 then IntegerVal (unsafeGetInteger obj)
+      else if prim_isVector obj == 1 then Vector (unsafeGetInteger (unsafeVectorRef obj 0))
+                                                 (readVector (unsafeVectorLength obj) 1 obj)
+      else if prim_isPair obj == 1 then Cons (decodeObj (unsafeFst obj))
+                                             (decodeObj (unsafeSnd obj))
+      else if prim_isFloat obj == 1 then FloatVal (unsafeGetFloat obj)
+      else if prim_isString obj == 1 then StringVal (unsafeGetString obj)
+      else if prim_isChar obj == 1 then CharVal (unsafeGetChar obj)
+      else if prim_isSymbol obj == 1 then Symbol (unsafeReadSymbol obj)
+      else if prim_isProcedure obj == 1 then Procedure obj
+      else if prim_isBox obj == 1 then Box (decodeObj (unsafeUnbox obj))
+      else Null
```

Comment:
I wonder if there's a better way to express this. I generally think of long conditionals as a code smell. 

Not to mention you have essentially the same conditional in the `WithTimeout` variant below. Could the runtime type checks be be folded into the `FromScheme` interface implementation, so that you only need to call `fromScheme` here?

Reply:
@emdash 

Still have this implemented this way, but am open to suggestions should there be a cleaner alternative.